llvm/lib/RemoteCachingService/CAS/GRPCRelayCAS.cpp

```diff
@@ -0,0 +1,504 @@
+
+  // Send the request to remote to load the object. Since Generator function
+  // can't return nullptr to indicate failed load now, we can't lock the entry
+  // into busy state. we will send parallel loads and put one of the result into
+  // the local CAS.
+  auto &GetQueue = CASDB->getQueue();
+  GetQueue.getAsync(getDataIDFromRef(Ref));
+  auto Response = GetQueue.receiveNext();
+  if (!Response)
+    return Response.takeError();
+
+  SmallVector<const InMemoryIndexValueT *> InternalRefs;
+  for (auto &ID : Response->Refs) {
+    auto &Ref = indexHash(arrayRefFromStringRef(ID));
+    InternalRefs.push_back(&Ref);
+  }
+  if (!Response->BlobData)
+    return createStringError(std::make_error_code(std::errc::invalid_argument),
+                             "Expect result to be returned in buffer");
+
+  ArrayRef<char> Content(Response->BlobData->data(),
+                         Response->BlobData->size());
+  return getObjectHandle(storeObjectImpl(I, InternalRefs, Content));
+}
+
+uint64_t GRPCRelayCAS::getDataSize(ObjectHandle Handle) const {
+  return getInMemoryCASData(Handle).getDataSize();
+}
+
+Error GRPCRelayCAS::forEachRef(ObjectHandle Handle,
+                               function_ref<Error(ObjectRef)> Callback) const {
+  auto &Node = getInMemoryCASData(Handle);
+  for (const InMemoryIndexValueT *Ref : Node.getRefs())
+    if (Error E = Callback(toReference(*Ref)))
+      return E;
+  return Error::success();
+}
+
+ObjectRef GRPCRelayCAS::readRef(ObjectHandle Node, size_t I) const {
+  return toReference(*getInMemoryCASData(Node).getRefs()[I]);
+}
+
+size_t GRPCRelayCAS::getNumRefs(ObjectHandle Handle) const {
+  return getInMemoryCASData(Handle).getNumRefs();
+}
+
+ArrayRef<char> GRPCRelayCAS::getData(ObjectHandle Handle,
+                                     bool RequiresNullTerminator) const {
+  return getInMemoryCASData(Handle).getData();
+}
+
+Expected<ObjectRef>
+GRPCRelayCAS::storeFromOpenFileImpl(sys::fs::file_t FD,
+                                    Optional<sys::fs::file_status> FS) {
+  std::error_code EC;
+  sys::fs::mapped_file_region Map(FD, sys::fs::mapped_file_region::readonly,
+                                  FS->getSize(),
+                                  /*offset=*/0, EC);
+  if (EC)
+    return errorCodeToError(EC);
+
+  auto &SaveQueue = CASDB->saveQueue();
+  ArrayRef<char> Data(Map.data(), Map.size());
+  SmallString<128> Path;
+  if (sys::fs::getRealPathFromHandle(FD, Path))
+    SaveQueue.saveFileAsync(Path.str().str());
+  else
+    SaveQueue.saveDataAsync(toStringRef(Data).str());
+
+  auto Response = SaveQueue.receiveNext();
+  if (!Response)
+    return Response.takeError();
+
+  auto &I = indexHash(arrayRefFromStringRef(Response->CASID));
+  // TODO: we can avoid the copy by implementing InMemoryRef object like
+  // InMemoryCAS.
+  return toReference(storeObjectImpl(I, None, Data));
+}
+
+Expected<ObjectRef> GRPCRelayCAS::storeFile(StringRef Content) {
```

Comment:
This is currently dead code. I can remove it since it is to test Load and Save GRPC protocol, which is currently used. Maybe it should just be dropped if CAS is not using them.